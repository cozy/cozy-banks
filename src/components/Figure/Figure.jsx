import React from 'react'
import Types from 'prop-types'
import cx from 'classnames'
import styles from './Figure.styl'

const currencySigns = {
  EUR: 'â‚¬',
  USD: '$'
}

/**
 * Shows a number, typically a balance or an important financial
 * number in a bold way.
 */
const stylePositive = styles['Figure-content--positive']
const styleNegative = styles['Figure-content--negative']
const styleWarning = styles['Figure-content--warning']
const styleBig = styles['Figure--big']
const Figure = props => {
  const {
    currency,
    coloredPositive,
    coloredNegative,
    coloredWarning,
    warningLimit,
    signed,
    className,
    total,
    size
  } = props

  let { decimalNumbers } = props
  decimalNumbers = isNaN(decimalNumbers) ? 2 : decimalNumbers

  const totalLocalized = total.toLocaleString('fr-FR', {
    minimumFractionDigits: decimalNumbers,
    maximumFractionDigits: decimalNumbers
  })
  const isTotalPositive = total > 0
  const isTotalInLimit = total > warningLimit
  const isWarning = !isTotalPositive && !isTotalInLimit && coloredWarning

  return (
    <div
      className={cx(
        {
          [stylePositive]: isTotalPositive && coloredPositive,
          [styleNegative]:
            total !== 0 && !isTotalPositive && !isWarning && coloredNegative,
          [styleWarning]: isWarning,
          [styleBig]: size == 'big'
        },
        className
      )}
    >
      <span className={styles['Figure-total']}>
        {isTotalPositive && signed && '+'}
        {totalLocalized}
      </span>
      {currency && (
        <span className={styles['Figure-currency']}>
          {currencySigns[currency] || currency}
        </span>
      )}
    </div>
  )
}

Figure.propTypes = {
  /** Number to display */
  total: Types.number.isRequired,
  /** Currency to show */
  currency: Types.string,
  /** Colors positive numbers in green */
  coloredPositive: Types.bool,
  /** Colors negative numbers in red */
  coloredNegative: Types.bool,
  /** Displays the sign */
  signed: Types.bool,
  /** Numbers of decimals to show (default=2) */
  decimalNumbers: Types.number,
  /** Whether to add a specific class to show warning */
  warningLimit: Types.number
}

export default Figure
