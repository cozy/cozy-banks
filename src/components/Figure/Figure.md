```jsx
<div>
  <Figure
    total={1000}
    symbol='€'
    coloredPositive coloredNegative signed />

  <Figure
    total={-1000}
    symbol='€'
    coloredPositive coloredNegative signed />

  <Figure
    total={-1000}
    symbol='€'
    signed />
</div>
```
