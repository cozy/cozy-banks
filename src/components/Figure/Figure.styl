@require 'settings/breakpoints'

.Figure-total
    font-weight    900

.Figure-currency
    margin-left .2em
    color       coolGrey

.Figure-content--positive
    color   emerald

    .Figure-currency
        color   emerald

.Figure-content--negative
    color   pomegranate

    .Figure-currency
        color   pomegranate

.Figure-content--warning
    color   texasRose

    .Figure-currency
        color   texasRose

.Figure--big
    font-size 2rem
    line-height 2.625rem

+small-screen()
    .Figure--big
        font-size 1.5rem
        line-height 1.75rem
