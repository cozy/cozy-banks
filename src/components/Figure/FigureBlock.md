Pour montrer une KPI importante.

```jsx
<div>
  <FigureBlock
    label='Balance totale'
    total={1000}
    symbol='€'
    coloredPositive coloredNegative signed />

  <FigureBlock
    label='Balance totale (negative number)'
    total={-1000}
    symbol='€'
    coloredPositive coloredNegative signed />

  <FigureBlock
    label='Balance totale (no color)'
    total={-1000}
    symbol='€'
    signed />
</div>
```
