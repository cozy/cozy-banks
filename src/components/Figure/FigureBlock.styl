@require 'settings/breakpoints'

.FigureBlock
    color charcoalGrey

.FigureBlock-label
    color coolGrey
    margin 0
    text-transform uppercase
    font-size 0.75rem
    line-height 1rem

.FigureBlock-figure
    font-size 2rem
    line-height 2.625rem

+small-screen()
    .FigureBlock
        font-size .7em

    .FigureBlock-figure
        padding-bottom 0.5rem
