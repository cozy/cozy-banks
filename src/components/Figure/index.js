export { default as Figure } from './Figure'
export { default as FigureBlock } from './FigureBlock'
