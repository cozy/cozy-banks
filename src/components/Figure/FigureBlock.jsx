import React from 'react'
import Types from 'prop-types'
import classNames from 'classnames'
import Figure from './Figure'
import styles from './FigureBlock.styl'

/**
 * Shows a `Figure` with a label.
 *
 * A part from `className` and `label`, takes same properties
 * as `Figure`.
 */
const FigureBlock = ({
  className,
  label,
  total,
  currency,
  coloredPositive,
  coloredNegative,
  signed,
  decimalNumbers = 0
}) => (
  <div className={classNames(styles['FigureBlock'], className)}>
    <h4 className={styles['FigureBlock-label']}>{label}</h4>
    <Figure
      size="big"
      className={styles['FigureBlock-figure']}
      total={total}
      currency={currency}
      coloredPositive={coloredPositive}
      coloredNegative={coloredNegative}
      signed={signed}
      decimalNumbers={decimalNumbers}
    />
  </div>
)

FigureBlock.propTypes = {
  /** Label of the figure */
  label: Types.string.isRequired,
  /** Extra classname */
  className: Types.string
}

export default FigureBlock
