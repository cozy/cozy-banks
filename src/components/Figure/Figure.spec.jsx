import React from 'react'
import { configure, shallow } from 'enzyme'
import flatten from 'lodash/flatten'
import merge from 'lodash/merge'
import Figure from './Figure'
import Adapter from 'enzyme-adapter-react-15'

configure({ adapter: new Adapter() })

const combine = (...iterables) => {
  if (iterables.length === 1) {
    return iterables[0].map(x => [x])
  } else {
    const combinationsNMinus1 = combine.apply(null, iterables.slice(1))
    return flatten(
      combinationsNMinus1.map(c => iterables[0].map(item => [item, ...c]))
    )
  }
}

const formatAttrs = attrs => {
  return Object.keys(attrs)
    .map(x => `${x}: ${attrs[x]}`)
    .join(', ')
}

describe('Figure', () => {
  const amounts = [-100, 100, 500, 4]

  const coloredAttributes = [
    'coloredPositive',
    'coloredNegative',
    'coloredWarning'
  ]

  const combinations = combine
    .apply(null, coloredAttributes.map(x => [{ [x]: true }, { [x]: false }]))
    .map(attrs => {
      return merge.apply(null, [{}, ...attrs])
    })

  for (let amount of amounts) {
    for (let attrs of combinations) {
      it(`should render correctly ${amount} ${formatAttrs(attrs)}`, () => {
        const el = shallow(
          <Figure warningLimit={110} total={amount} {...attrs} />
        ).getElement()
        expect(el).toMatchSnapshot()
      })
    }
  }
})
