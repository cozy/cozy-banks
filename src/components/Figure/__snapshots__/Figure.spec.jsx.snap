// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Figure should render correctly -100 coloredPositive: false, coloredNegative: false, coloredWarning: false 1`] = `
<div
  className=""
>
  <span
    className="Figure-total"
  >
    -100.00
  </span>
</div>
`;

exports[`Figure should render correctly -100 coloredPositive: false, coloredNegative: false, coloredWarning: true 1`] = `
<div
  className="Figure-content--warning"
>
  <span
    className="Figure-total"
  >
    -100.00
  </span>
</div>
`;

exports[`Figure should render correctly -100 coloredPositive: false, coloredNegative: true, coloredWarning: false 1`] = `
<div
  className="Figure-content--negative"
>
  <span
    className="Figure-total"
  >
    -100.00
  </span>
</div>
`;

exports[`Figure should render correctly -100 coloredPositive: false, coloredNegative: true, coloredWarning: true 1`] = `
<div
  className="Figure-content--warning"
>
  <span
    className="Figure-total"
  >
    -100.00
  </span>
</div>
`;

exports[`Figure should render correctly -100 coloredPositive: true, coloredNegative: false, coloredWarning: false 1`] = `
<div
  className=""
>
  <span
    className="Figure-total"
  >
    -100.00
  </span>
</div>
`;

exports[`Figure should render correctly -100 coloredPositive: true, coloredNegative: false, coloredWarning: true 1`] = `
<div
  className="Figure-content--warning"
>
  <span
    className="Figure-total"
  >
    -100.00
  </span>
</div>
`;

exports[`Figure should render correctly -100 coloredPositive: true, coloredNegative: true, coloredWarning: false 1`] = `
<div
  className="Figure-content--negative"
>
  <span
    className="Figure-total"
  >
    -100.00
  </span>
</div>
`;

exports[`Figure should render correctly -100 coloredPositive: true, coloredNegative: true, coloredWarning: true 1`] = `
<div
  className="Figure-content--warning"
>
  <span
    className="Figure-total"
  >
    -100.00
  </span>
</div>
`;

exports[`Figure should render correctly 4 coloredPositive: false, coloredNegative: false, coloredWarning: false 1`] = `
<div
  className=""
>
  <span
    className="Figure-total"
  >
    4.00
  </span>
</div>
`;

exports[`Figure should render correctly 4 coloredPositive: false, coloredNegative: false, coloredWarning: true 1`] = `
<div
  className=""
>
  <span
    className="Figure-total"
  >
    4.00
  </span>
</div>
`;

exports[`Figure should render correctly 4 coloredPositive: false, coloredNegative: true, coloredWarning: false 1`] = `
<div
  className=""
>
  <span
    className="Figure-total"
  >
    4.00
  </span>
</div>
`;

exports[`Figure should render correctly 4 coloredPositive: false, coloredNegative: true, coloredWarning: true 1`] = `
<div
  className=""
>
  <span
    className="Figure-total"
  >
    4.00
  </span>
</div>
`;

exports[`Figure should render correctly 4 coloredPositive: true, coloredNegative: false, coloredWarning: false 1`] = `
<div
  className="Figure-content--positive"
>
  <span
    className="Figure-total"
  >
    4.00
  </span>
</div>
`;

exports[`Figure should render correctly 4 coloredPositive: true, coloredNegative: false, coloredWarning: true 1`] = `
<div
  className="Figure-content--positive"
>
  <span
    className="Figure-total"
  >
    4.00
  </span>
</div>
`;

exports[`Figure should render correctly 4 coloredPositive: true, coloredNegative: true, coloredWarning: false 1`] = `
<div
  className="Figure-content--positive"
>
  <span
    className="Figure-total"
  >
    4.00
  </span>
</div>
`;

exports[`Figure should render correctly 4 coloredPositive: true, coloredNegative: true, coloredWarning: true 1`] = `
<div
  className="Figure-content--positive"
>
  <span
    className="Figure-total"
  >
    4.00
  </span>
</div>
`;

exports[`Figure should render correctly 100 coloredPositive: false, coloredNegative: false, coloredWarning: false 1`] = `
<div
  className=""
>
  <span
    className="Figure-total"
  >
    100.00
  </span>
</div>
`;

exports[`Figure should render correctly 100 coloredPositive: false, coloredNegative: false, coloredWarning: true 1`] = `
<div
  className=""
>
  <span
    className="Figure-total"
  >
    100.00
  </span>
</div>
`;

exports[`Figure should render correctly 100 coloredPositive: false, coloredNegative: true, coloredWarning: false 1`] = `
<div
  className=""
>
  <span
    className="Figure-total"
  >
    100.00
  </span>
</div>
`;

exports[`Figure should render correctly 100 coloredPositive: false, coloredNegative: true, coloredWarning: true 1`] = `
<div
  className=""
>
  <span
    className="Figure-total"
  >
    100.00
  </span>
</div>
`;

exports[`Figure should render correctly 100 coloredPositive: true, coloredNegative: false, coloredWarning: false 1`] = `
<div
  className="Figure-content--positive"
>
  <span
    className="Figure-total"
  >
    100.00
  </span>
</div>
`;

exports[`Figure should render correctly 100 coloredPositive: true, coloredNegative: false, coloredWarning: true 1`] = `
<div
  className="Figure-content--positive"
>
  <span
    className="Figure-total"
  >
    100.00
  </span>
</div>
`;

exports[`Figure should render correctly 100 coloredPositive: true, coloredNegative: true, coloredWarning: false 1`] = `
<div
  className="Figure-content--positive"
>
  <span
    className="Figure-total"
  >
    100.00
  </span>
</div>
`;

exports[`Figure should render correctly 100 coloredPositive: true, coloredNegative: true, coloredWarning: true 1`] = `
<div
  className="Figure-content--positive"
>
  <span
    className="Figure-total"
  >
    100.00
  </span>
</div>
`;

exports[`Figure should render correctly 500 coloredPositive: false, coloredNegative: false, coloredWarning: false 1`] = `
<div
  className=""
>
  <span
    className="Figure-total"
  >
    500.00
  </span>
</div>
`;

exports[`Figure should render correctly 500 coloredPositive: false, coloredNegative: false, coloredWarning: true 1`] = `
<div
  className=""
>
  <span
    className="Figure-total"
  >
    500.00
  </span>
</div>
`;

exports[`Figure should render correctly 500 coloredPositive: false, coloredNegative: true, coloredWarning: false 1`] = `
<div
  className=""
>
  <span
    className="Figure-total"
  >
    500.00
  </span>
</div>
`;

exports[`Figure should render correctly 500 coloredPositive: false, coloredNegative: true, coloredWarning: true 1`] = `
<div
  className=""
>
  <span
    className="Figure-total"
  >
    500.00
  </span>
</div>
`;

exports[`Figure should render correctly 500 coloredPositive: true, coloredNegative: false, coloredWarning: false 1`] = `
<div
  className="Figure-content--positive"
>
  <span
    className="Figure-total"
  >
    500.00
  </span>
</div>
`;

exports[`Figure should render correctly 500 coloredPositive: true, coloredNegative: false, coloredWarning: true 1`] = `
<div
  className="Figure-content--positive"
>
  <span
    className="Figure-total"
  >
    500.00
  </span>
</div>
`;

exports[`Figure should render correctly 500 coloredPositive: true, coloredNegative: true, coloredWarning: false 1`] = `
<div
  className="Figure-content--positive"
>
  <span
    className="Figure-total"
  >
    500.00
  </span>
</div>
`;

exports[`Figure should render correctly 500 coloredPositive: true, coloredNegative: true, coloredWarning: true 1`] = `
<div
  className="Figure-content--positive"
>
  <span
    className="Figure-total"
  >
    500.00
  </span>
</div>
`;
